src/librustc_back/target/mod.rs

```diff
@@ -391,7 +397,13 @@ impl Default for TargetOptions {
             allow_asm: true,
             has_elf_tls: false,
             obj_is_bitcode: false,
-            max_atomic_width: 0,
+            // NOTE this is *not* the real default value. The default value of max_atomic_width is
+            // actually the pointer width of the target. This default is injected in the
+            // Target::from_json function. Still, we have to pick some value to put here. We'll pick
+            // an impossible value: u64::max_value() because using a valid value like 0 or 8 as the
+            // default would cause the max-atomic-width field to be "lost" (it would get replaced
+            // by target_pointer_width) during the Target <-> JSON round trip
+            max_atomic_width: u64::max_value(),
```

Comment:
Sure! Let me ellaborate here, first, what happens today without this change:

- All targets go through a round trip from `TargetOptions` to JSON and back to `TargetOptions`.
- Let's say you set `max_atomic_width` to the `Default` value, which is `0` today, for a target specification, when this specification is converted to JSON the `max_atomic_width` field is *not* present in that JSON representation.
- When that JSON representation is converted back to `TargetOptions`, the current logic sets `max_atomic_width` to `target_pointer_witdh`, i.e. `32` or `64`, because the `max_atomic_width` field is not present in the JSON representation.

TL;DR You set `max_atomic_witdth` to `0` but the compiler is actually using `32` or `64` for the target.

The bug is subtle and the proposed fix is sort of magical. Perhaps we could solve it in a better way.